core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -233,7 +234,18 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
         replicaManager.startup()
 
         /* start kafka controller */
-        kafkaController = new KafkaController(config, zkUtils, time, metrics, threadNamePrefix)
+        val zookeeperClient = new ZookeeperClient(config.zkConnect, config.zkSessionTimeoutMs,
+          config.zkConnectionTimeoutMs, new StateChangeHandler {
+            override def onConnectionTimeout = ???
+
+            override def afterInitializingSession = kafkaController.newSession()
+
+            override def onAuthFailure = ???
+
+            override def beforeInitializingSession = kafkaController.expire()
+          })
+        kafkaControllerZkUtils = new KafkaControllerZkUtils(zookeeperClient, zkUtils.isSecure)
+        kafkaController = new KafkaController(config, kafkaControllerZkUtils, time, metrics, threadNamePrefix)
```

Comment:
We probably want to wait for  ZookeeperClient to be connected within the connection timeout. Otherwise, fail the restart of the broker.

Reply:
This is already exactly what happens. The ZookeeperClient constructor calls `waitUntilConnected(connectionTimeoutMs, TimeUnit.MILLISECONDS)`.